### PR TITLE
[ios] Work around annotation race condition

### PR DESCRIPTION
**Pull request by boundsj**
Fixes https://github.com/mapbox/mapbox-gl-native/issues/6901

When annotations are removed concurrently with being updated because of a map redraw, it is possible that mbgl will still consider an annotation to exist depending on what part of the map lifecycle the call update annotations happens. However, the annotation context by
tag map in MGLMapView (on the main thread) will always be up to date.
This adds a guard that uses that map to avoid considering annotations that have actually been removed as visible.

Another issue was that adding the same annotation object twice has always been allowed and has worked, visually. However, a recent refactor to make the reuse queue work correctly is not tolerant of the a single view representing two annotation instances that are the same.
The effect is that a new UIView will be generated for each (clone) annotation that is added and only the last view will move with the map. The other views become detached and float around in the view as the map is panned. This commit adds a guard to make adding an annotation instance a no-op if it has already been added before.

cc @1ec5 @incanus @bsudekum

### CODE REVIEW

File: platform/ios/src/MGLMapView.mm [hunk 1]
```diff
@@ -2836,7 +2836,11 @@ - (void)removeStyleClass:(NSString *)styleClass
         
         for (auto const& annotationTag: annotationTags)
         {
-            MGLAnnotationContext annotationContext = _annotationContextsByAnnotationTag[annotationTag];
+            if (!_annotationContextsByAnnotationTag.count(annotationTag))
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: Can you make these changes in the macOS implementation too?

Comment (boundsj) [inline platform/ios/src/MGLMapView.mm]: Oops. Done in 02e5978

File: platform/ios/src/MGLMapView.mm [hunk 2]
```diff
@@ -2903,6 +2907,12 @@ - (void)addAnnotations:(NS_ARRAY_OF(id <MGLAnnotation>) *)annotations
     {
         NSAssert([annotation conformsToProtocol:@protocol(MGLAnnotation)], @"annotation should conform to MGLAnnotation");
 
+        // adding the same annotation object twice is a no-op
+        if ([self.annotations containsObject:annotation])
```
Comment (1ec5) [inline platform/ios/src/MGLMapView.mm]: Also make this change in the macOS implementation.

Comment (boundsj) [inline platform/ios/src/MGLMapView.mm]: 😬  done in ecf14e3

**Pull request merged**